### PR TITLE
Cass client pool refactor

### PR DESCRIPTION
**Pull request by sudiksha27**
**Goals (and why)**:
The `CassandraClientPool` needs refactoring in order to be functional with the Cassandra server deployed in k8s infra.
This PR lays the groundwork of this refactor. 

**Implementation Description (bullets)**:
We introduce a new model `CassandraServer` which contains the actual cassandraHost Ip and the proxy Ip which we will use to talk the cassandraHost. 
In this PR all of the comparisons (for blacklisting, absentHostTracker etc) are done on the cassandraHost returned by the `describe_ring` endpoint. The actual call to Cassandra are made using a proxy.

**Testing (What was existing testing like?  What have you done to improve it?)**:
Relying on existing tests for sanity test. We are designing a sensible e2e test on our internal dev stack.

**Concerns (what feedback would you like?)**:
Does the design look sensible? 
Did I destroy an invariant?

**Where should we start reviewing?**:
CassandraServer, CassandraService

**Priority (whenever / two weeks / yesterday)**:
yesterday 🏃🏼‍♀️

### CODE REVIEW

**Comment by sudiksha27** [comment]
I do not really want to merge the PR as it is. I am raising it for the easy of reviewing. I would be cutting another branch from this which will incorporate changes to the way `CassandraServer` will be used.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolingContainer.java
Comment (sudiksha27) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolingContainer.java:77]: cassandraServer is only used for logging. You can see that the impl of `CassandraClientFactory` has not changed. We pass in the proxy field to `CassandraClientFactory` that creates a client.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/Blacklist.java
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/Blacklist.java:145]: only called from trace log is we cannot find targeted host for key

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java:397]: It is a one time start up check, even if too long we fail and do not start up so ....the worst has already happened.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java:511]: `CassandraClientFactory` basically has not changed at all. Maybe sneaky in the way that we are hard binding a proxy Ip to a cass server ip but we will be changing this algorithm anyway. We can also change `CassandraClientFactory` to look uniform later.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java:314]: hmm we are only comparing the cass hosts above - so reaching this point would mean that we saw a new cass host in the describe_ring op.
okay i will move that super noisy log line here and log conditionally if the proxies do not match.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java:210]: only added for RC - will resolve this comment as soon as i cut an RC

**Review by jeremyk-91 — APPROVED** [review]
Bunch of small things, but I'm fairly convinced that this works

**Comment by jeremyk-91** [comment]
note:
- up to this point, we populate the full list of proxies in `CassandraServer`, but only expose `proxy[0]` to the rest of the underlying code
- this is assumed to be stable
- and the check for absent etc. is based on `CassandraServer` (with some maybe noisy logging for when the server is the same one but the proxies don't match, but we're looking to validate that on a development stack).

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/HostPartitioner.java
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/HostPartitioner.java]: Yeah basically I had to change the apis to use `CassServer`, the todo was not required

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/Blacklist.java
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/Blacklist.java:129]: great catch, esp in this PR

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java:315]: The tests were failing over something else. Why does this one look sus?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java
Comment (gsheasby) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java:397]: 🔥 ok then!

File: atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolTest.java
Comment (gsheasby) [inline atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolTest.java:476]: nit: clientPoolWithProxies maybe?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java
Comment (gsheasby) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java:315]: It was just non-trivial to trace where these methods were coming from, and the logic in this return statement in particular is quite magical.

The `currentPools.keySet().stream().map(CassandraServer::proxy).collect(Collectors.toSet())` part could possibly be extracted into `getProxiesFromCurrentPools()`.

Also, isn't `config.servers().accept(new ThriftHostExtractingVisitor()` already a method somewhere? Or has that gone now?

With that, we'd just have `return ImmutableSet.copyOf(Sets.union(getProxiesFromCurrentPools(), getHostsFromConfig())`, which is easier to understand.

**Review by gsheasby — APPROVED** [review]
LGTM, obviously pending:
1 - testing (!)
2 - removing or limiting the potentially noisy log lines

**Review by Jolyon-S — CHANGES_REQUESTED** [review]
I think there are a couple of places where we're recomputing the same DNS lookups for a given IP address. I don't think this affects correctness, but we can probably just do better. Still reviewing the rest.

[EDIT] Or maybe not, but it's likely that we can refactor the code to make this clearer!

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java:434]: nit: is it a host or a server?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java:551]: this seems like it'll be very big - will that cause memory issues?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServer.java
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServer.java:31]: this needs rewording.
```suggestion
     * We maintain set of all IPs for this server, but do not create a client pool for each one.
```

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServer.java
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServer.java:41]:
```suggestion
        return reachableProxyIps().stream().findFirst().orElseThrow();
```

Comment (gsheasby) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServer.java:41]: Also, how about adding a `@Value.Check` to disallow setting `reachableProxyIps` to the empty set?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java:195]: What happens if config does not? Do we just crash and burn?

Can we detect this case, and if so, can we log something useful?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java:289]: what does the "throw unchecked" mean here?

File: atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolTest.java
Comment (gsheasby) [inline atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolTest.java:163]: nit: perhaps extract a method here that you can then use in WeightedServersTest

**Review by gsheasby — COMMENTED** [review]
Generally looks good, I have a few nits, and there's one piece in particular that needs to be reworded.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java:551]: _Maybe_ but at this point we are in an inconsistent state and should not be writing to Cass anyway

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServer.java
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServer.java:41]: we have that :)

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java:195]: well we will not be able to talk to Cassandra at all, think those are logged as somewhat meaningful messages already

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java:289]: err i killed the other method so this makes no sense now, will rename

File: atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolTest.java
Comment (gsheasby) [inline atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolTest.java:218]: I meant to make this (public/package) static, and then use in other places where you're doing the `"0", 0` thing - but certainly not blocking on this! Fine to test/merge as-is.

**Review by gsheasby — APPROVED** [review]

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServer.java
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServer.java:30]: form :)

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java:270]: Does this make sense now that we are moving to Rubix?

Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java:270]: plus we have the persisted hostIp -> hostname table

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraAbsentHostTracker.java
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraAbsentHostTracker.java:38]: Still have the situation where pod in k8s changes IP upon bounce - I think it is useful to have the absent host tracker

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java
Comment (sudiksha27) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java:302]: inlined

**Comment by jeremyk-91** [comment]
pending smoketest, fine to go in current state 👍

**Pull request merged**
